suite/tests/client-interface/drbbdup-meta-nop-bb-test.dll.c

```diff
@@ -193,17 +193,24 @@ instrument_instr(void *drcontext, void *tag, instrlist_t *bb, instr_t *instr,
     if (is_cur_all_meta)
         CHECK(is_meta_nop(instr), "should be the meta nop instruction");
 
+    bool has_emul_labels = drmgr_is_emulation_start(instr) || instr_is_syscall(instr) ||
```

Comment:
I do not get these changes. Why would emulation affect results of drbbdup_is_first_nonlabel_instr, drbbdup_is_first_instr, etc, and therefore we need to edge them out in these test-cases? Should drbbdup internally  handle the emulation of the jmp exit instruction? From the perspective of the client, the jmp exit instruction is invisible.

Reply:
The emulation labels are passed to the insert-instrumentation callbacks: so this code which expects just one instruction sees 3 instructions for the syscall in the asm: label;syscall;label.

Reply:
Why should drbbdup pass them to the client? I guess this is an indicator for client to call drmgr_orig_app_instr_for_fetch(), but can't the client just call it anyways and check for nullness?

Reply:
Note that drbbdup_is_last_instr() should return true when the special instruction is passed. If I pass the emulated instruction to drbbdup_is_last_instr(), does it return true?

Reply:
OK you are saying can drbbdup hide the emulation markers from the client for the drbbdup-controlled instru insertion callbacks, and only have drmgr see them for the drmgr emulation queries.  Yes that seems doable.

Reply:
And the pt->last_instr needs to skip over the emulation labels.  Done.